Mage.Server.Plugins/Mage.Player.Human/src/mage/player/human/HumanPlayer.java

```diff
@@ -116,8 +123,36 @@ public HumanPlayer(final HumanPlayer player) {
         this.currentlyUnpaidMana = player.currentlyUnpaidMana;
         this.replacementEffectChoice = player.replacementEffectChoice;
     }
+    
+    protected boolean isExecutingMacro() {
+        return !actionQueue.isEmpty() || (actionIterations > 0 && !actionQueueSaved.isEmpty());
+    }
+    
+    protected boolean pullResponseFromQueue(Game game) {
+        if (actionQueue.isEmpty() && actionIterations > 0 && !actionQueueSaved.isEmpty()) {
+            actionQueue = new LinkedList(actionQueueSaved);
+            actionIterations--;
+        }
+        PlayerResponse action = actionQueue.poll();
+        if (action != null) {
+            if (action.getString() != null
+                && action.getString().equals("resolveStack")) {
+                action = actionQueue.poll();
+                if (action == null) return false;
+                sendPlayerAction(PlayerAction.PASS_PRIORITY_UNTIL_STACK_RESOLVED, game, null);
+            }
+            synchronized (response) {
+                response.copy(action);
+                response.notifyAll();
+                macroTriggeredSelectionFlag = false;
+                return true;
+            }
+        }
+        return false;
+    }
 
     protected void waitForResponse(Game game) {
+        if(!recordingMacro && pullResponseFromQueue(game)) return;
```

Comment:
Not a fan of these conditions and execution on the same line. Condition at least should go on a new line indented - it's hard to see code flow especially with the return inline here.